src/main/java/com/coreos/jetcd/EtcdWatchImpl.java

```diff
@@ -0,0 +1,294 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * etcd watcher Implementation
+ */
+public class EtcdWatchImpl implements EtcdWatch {
+
+    private StreamObserver<WatchRequest> requestStream;
+
+    private ConcurrentHashMap<Long, Watcher> watchers = new ConcurrentHashMap<>();
+
+    private WatchGrpc.WatchStub watchStub;
+
+    private ConcurrentLinkedQueue<Pair<Watcher.Builder, ListenableSetFuture<Watcher>>> pendingWatchers = new ConcurrentLinkedQueue<>();
+    private Map<Long, Watcher> cancelWatchers = new ConcurrentHashMap<>();
+
+    public EtcdWatchImpl(WatchGrpc.WatchStub watchStub) {
+        this.watchStub = watchStub;
+    }
+
+    /**
+     * Watch watches on a key or prefix. The watched events will be called by onWatch.
+     * If the watch is slow or the required rev is compacted, the watch request
+     * might be canceled from the server-side and the onCreateFailed will be called.
+     *
+     * @param key         the key subscribe
+     * @param watchOption key option
+     * @param callback    call back
+     * @return ListenableFuture watcher
+     */
+    @Override
+    public synchronized ListenableSetFuture<Watcher> watch(ByteString key, WatchOption watchOption, Watcher.WatchCallback callback) {
+        WatchRequest request = optionToWatchCreateRequest(key, watchOption);
+        Watcher.Builder builder = Watcher.newBuilder().withCallBack(callback)
+                .withKey(key)
+                .withWatchOption(watchOption);
+        ListenableSetFuture<Watcher> waitFuture = new ListenableSetFuture<>(null);
+        this.pendingWatchers.add(new Pair<>(builder, waitFuture));
+        getRequestStream().onNext(request);
+        return waitFuture;
+    }
+
+    /**
+     * Cancel the watch task with the watcher, the onCanceled will be called after successfully canceled.
+     *
+     * @param watcher the watcher to be canceled
+     */
+    @Override
+    public void cancelWatch(Watcher watcher) {
+        Watcher temp = watchers.get(watcher.getWatchID());
+        if (temp != null) {
+            synchronized (temp) {
+                if (this.watchers.containsKey(temp.getWatchID())) {
+                    this.watchers.remove(temp.getWatchID());
+                    this.cancelWatchers.put(temp.getWatchID(), temp);
+                    WatchCancelRequest cancelRequest = WatchCancelRequest.newBuilder().setWatchId(watcher.getWatchID()).build();
+                    WatchRequest request = WatchRequest.newBuilder().setCancelRequest(cancelRequest).build();
+                    this.requestStream.onNext(request);
+                }
+            }
+        }
+    }
+
+    /**
+     * empty the old request stream, watchers and resume the old watchers
+     * empty the cancelWatchers as there is no need to cancel, the old request stream has been dead
+     */
+    private void resume() {
+        synchronized (this) {
```

Comment:
since the whole code block is synchronized, you could make this method synchronized.